pkglistgen/tool.py

```diff
@@ -717,6 +720,14 @@ def update_and_solve_target(
         if not product_version:
             # for stagings the product version doesn't matter (I hope)
             product_version = '1'
+        with open(product_dir + "/release.changes", 'w', encoding="utf-8") as f:
+            date = package_source_changed(api.apiurl, project, group).replace(tzinfo=timezone.utc)
```

Comment:
IMO the changelog date should really state when the change was done and that's `datetime.now()`

Reply:
Why?

AFAIK it was not done now, only the build was done now. It may also have been rebuilt resulting in no change if the input was the same. So I think it was done when the source (the group package) was commited (or earlier). I feel it is similar to that we don't change the changelog when we rebuild rpms. I do think we gain no information by storing when it was built.

The idea of reproducible builds is that the build date doesn't matter as the output will be the same no matter at what time the build is done. The advantage of reproducible builds is that to security review you only review the source and not the build output, to verify the build output you only need to bit wise compare it to a rebuild. Embedding $now would prevent that.

Reply:
https://github.com/openSUSE/openSUSE-release-tools/pull/3062#issuecomment-1956449655: Only if a change was actually performed, add a changes entry which contains `datetime.now()`.

`package_source_changed(api.apiurl, project, group)` is constant except for manual changes, so the current code will end up using a backdated changelog entry, resulting in mtimes which don't really make sense.

Reply:
Mh. Instead of just the date any input to the release package changed, a different date that seems useful to me is when any source package in the whole distro last changed. Sadly that is probably too work intensive to compute with OBS-vcs, but it hopefully will be easy when we move to git.

Your suggestion to only change the date when the output changes, but then to $now does not make sense to me, because that is neither of those and non-reproducible (by the definition of https://reproducible-builds.org/ ).

Any date is always sort of backdated due to latency until being observed. Latency from one CPU sending to it being observed, e.g. by an output device like a disk or another CPU core in a multi core machine. Time is not intuitive like you imply. So IMHO the date being backdated, doesn't make it make less sense, even when it is further back by magnitudes.

It seems to me that it being reproducible and knowing when a human with osc commit triggered the change and defined the full outcome of the computation has more value than knowing what clock the CPU perceived when it did the computation.

I don't know of any case where using $now here would actually break something, it just hinders making this build step reproducible, but at least not the next ones.

If you prefer I can change it to always $now and when we move to any distro from obs-vcs to git, we can discuss again.

Reply:
See https://github.com/openSUSE/openSUSE-release-tools/pull/3064 for the alternative with $now.

Reply:
> Your suggestion to only change the date when the output changes, but then to $now does not make sense to me, because that is neither of those and non-reproducible (by the definition of https://reproducible-builds.org/ ).

It is perfectly reproducible as the date is saved. When I do `osc vc`, the same happens and it's fine.

Reply:
There I'm not talkling about the build output of the release rpm, but the source of it, which is the build output of this script.

Reply:
As pointed out in https://github.com/openSUSE/openSUSE-release-tools/pull/3064#issuecomment-1985443655 I missed that some of the input is from 000update-repos . I added that. It seems 000product is not an input. Did I miss anything else?

Reply:
the binary list of the FTP tree is part of the input. Any name that disappears is added to the weakremovers.inc

Reply:
ah it gets that from the solv modules. *sigh*. Then there is no easy way around it. Thank you.